### ISSUE TITLE
Spurious unused value warning with destructive substitution

### ISSUE THREAD

**Issue by vicuna** [ocaml/ocaml#7852]
**Original bug ID:** 7852
**Reporter:** @trefis
**Assigned to:** @trefis
**Status:** assigned (set by @trefis on 2018-09-21T16:46:49Z)
**Resolution:** open
**Priority:** normal
**Severity:** minor
**Version:** 4.08.0+dev/beta1/beta2
**Category:** typing
**Monitored by:** @nojb @hhugo @Drup

## Bug description

This regression was introduced by https://github.com/ocaml/ocaml/pull/1737

I'll try to fix it without having to revert the GPR.
But if I fail, we can always revert it, it's not that important.

## Steps to reproduce

```
$ cat test.mli
module M : sig
  type t
  val foo : t -> int
  val bar : t -> int
end

module N : sig
  type outer
  type t
  val foo : t -> outer
  val bar : t -> outer
end with type outer := int
$ ./ocamlc.opt -w +32 -c ./test.mli
File "./test.mli", line 10, characters 2-22:
Warning 32: unused value foo.
File "./test.mli", line 11, characters 2-22:
Warning 32: unused value bar.
$ $HOME/.opam/4.07.0/bin/ocamlc -w +32 -c ./test.mli
$
```

**Comment by vicuna** [comment]
**Comment author:** @trefis

Btw, this would also happen when destructively substituting a module.

**Comment by vicuna** [comment]
**Comment author:** @trefis

This was reverted in https://github.com/ocaml/ocaml/pull/2092

FTR, what was happening is basically this:

Usage of items in OCaml is tracked by their name (i.e. the string, not the ident) and a Location.
When first typing the "sig .. end" part of "sig .. end with ..." (that is done in Typemod), we enter the names and locations of foo and bar in some table in Env.
When we process the constraint (the "with ..." part) (also done in Typemod), we assign a new location to these items.

Once we're done with the typing (we're now in Compile_common.typecheck_intf), we call "ignore (Includemod.signatures info.env sg sg)". Which has the effect of marking every item in the signature as used (which makes sense for mlis).

The issue is that we're doing that by going over the signature, taking the name + location of every item, and then using that to look in the table in Env. But the location we have is the new one, not the initial one which was used to add the items to the table. So we never mark the items that were initially added.
And so we get the spurious warnings.

**Comment by vicuna** [comment]
**Comment author:** @gasche

Thanks for the summary, I get it now. (Another motivation to have a proper "declaration path" notion instead of source locations.)

**Comment by hhugo** [comment]
#1737 was reverted in #2092. Can be closed.

**Comment by trefis** [comment]
Actually, I’d like to keep this open until “a proper fix” is implemented.

**Comment by trefis** [comment]
> Another motivation to have a proper "declaration path" notion instead of source locations.

I wrote a prototype for that based on 4.07 last week (because at the time it's still wasn't easy to test 4.08 / trunk on actual code) which I plan to rebase onto trunk (and submit a PR) next week.

**Comment by gasche** [comment]
Thanks! I vaguely thought of this again when seeing the `error` type of includemod.ml while reviewing #8546:

```ocaml
type pos =
    Module of Ident.t | Modtype of Ident.t | Arg of Ident.t | Body of Ident.t
type error = pos list * Env.t * symptom
```

**Comment by lpw25** [comment]
@trefis Has this been fixed now?

**Comment by trefis** [comment]
The issue has been fixed by reverting #1737, and a couple of weeks ago the "proper fix" I was talking about got merged (#8934).
I think I only need to resubmit #1737 for inclusion before we can finally close this issue.

I'll try to do that in the not too distant future.

**Comment by trefis** [comment]
Actually, looks like #7859 is already tracking #1737 reinclusion, so let's close this.
